foyer/forcefield.py

```diff
@@ -947,5 +967,418 @@ def _write_references_to_file(self, atom_types, references_file):
+            ):
+                params = {
+                    "k": harmonic_bond_force_gen.k[i],
+                    "length": harmonic_bond_force_gen.length[i],
+                }
+                break
+
+        if params is None:
+            raise MissingParametersError(
+                f"Missing HarmonicBond parameters for bond between "
+                f"atoms {atom_1_type} and {atom_2_type}"
+            )
+
+        return params
+
+    def _extract_harmonic_angle_params(self, atom_types):
+        """Returns parameters for a HarmonicAngle between atom types"""
+        if len(atom_types) != 3:
+            raise ValueError(
+                f"HarmonicAngle parameters can only "
+                f"be extracted for three atoms. Provided {len(atom_types)}"
+            )
+
+        harmonic_angle_force_gen = self.get_generator(
+            ff=self, gen_type=HarmonicAngleGenerator
+        )
+
+        atom_1_type = atom_types[0]
+        atom_2_type = atom_types[1]
+        atom_3_type = atom_types[2]
+
+        angles_for_atom_2_type = harmonic_angle_force_gen.anglesForAtom2Type
+        params = None
+
+        for i in angles_for_atom_2_type[atom_2_type]:
+            types1 = harmonic_angle_force_gen.types1[i]
+            types2 = harmonic_angle_force_gen.types2[i]
+            types3 = harmonic_angle_force_gen.types3[i]
+
+            if (
+                atom_1_type in types1
+                and atom_2_type in types2
+                and atom_3_type in types3
+            ) or (
+                atom_3_type in types1
+                and atom_2_type in types2
+                and atom_1_type in types3
+            ):
+                params = {
+                    "theta": harmonic_angle_force_gen.angle[i],
+                    "k": harmonic_angle_force_gen.k[i],
+                }
+                break
+
+        if params is None:
+            raise MissingParametersError(
+                f"Missing HarmonicAngle parameters for angles between "
+                f"atoms {atom_1_type}, {atom_2_type} and {atom_3_type}"
+            )
+
+        return params
+
+    def _extract_periodic_proper_params(self, atom_types):
+        """Returns parameters for a PeriodicTorsion (proper) between atom types"""
+        if len(atom_types) != 4:
+            raise ValueError(
+                f"PeriodicTorsion proper parameters can only "
+                f"be extracted for four atoms. Provided {len(atom_types)}"
+            )
+
+        periodic_torsion_force_gen = self.get_generator(
+            ff=self, gen_type=PeriodicTorsionGenerator
+        )
+
+        wildcard = self._atomClasses[""]
+        (
+            atom_1_type,
+            atom_2_type,
+            atom_3_type,
+            atom_4_type,
+        ) = self.substitute_wildcards(atom_types, wildcard)
+
+        match = None
+        for index in periodic_torsion_force_gen.propersForAtomType[atom_2_type]:
```

Comment:
May be proper matching can also be consolidated into a single function. Needs further investigation.